### PR TITLE
psa: Define mbedtls_ecc_group_to_psa() inline

### LINKED ISSUE #3300: pk.c fails to link on dual world systems
### Description
- Type: Bug
- Priority: Major

---------------------------------------------------------------
## Bug

**OS**  
Mbed OS

**mbed TLS build:**  
Version: 2.22.0
Mbed OS version: 5.15.2

**Expected behavior**   
Linker works without errors when building TLS+PK on NS world, with psa_crypto.c on S world.

**Actual behavior**  
Linker generates errors when linking pk.c on NS.

[Error] @0,0: L6218E: Undefined symbol mbedtls_ecc_group_to_psa (referred from BUILD/LPC55S6
9_NS/ARM/mbed-os/features/mbedtls/mbed-crypto/src/pk.o)

**Steps to reproduce** 

Build and link pk.c separately from psa_crypto.c

### PR DESCRIPTION
**Pull request by Patater**
On dual world platforms, we want to run the PK module (pk.c) on the NS
side so TLS can use PSA APIs via the PK interface. PK currently has a
hard dependency on mbedtls_ecc_group_to_psa() which is declared in
crypto_extra.h, but only defined in psa_crypto.c, which is only built
for the S side.

Without this change, dual world platforms get error messages like the
following.

    [Error] @0,0: L6218E: Undefined symbol mbedtls_ecc_group_to_psa (referred from BUILD/LPC55S69_NS/ARM/mbed-os/features/mbedtls/mbed-crypto/src/pk.o)

Make mbedtls_ecc_group_to_psa() inline within crypto_extra.h so that it
is available to both NS and S world code.

Fixes #3300

Signed-off-by: Darryl Green <darryl.green@arm.com>
Signed-off-by: Jaeden Amero <jaeden.amero@arm.com>

## Status
**READY**

## Requires Backporting
NO  - PSA related change, and no LTS branches contain PSA.

## Migrations
mbedtls_ecc_group_to_psa() is becoming inline. This will require a rebuild of code using this function.

## Todos
- [X] Tests - covered by Mbed OS CI on LPC55S69
- ~[ ] Documentation~
- [X] Changelog updated
- ~[ ] Backported~

### CODE REVIEW

**Comment by mpg** [comment]
The CI's not very happy about this:
```
In file included from ../include/psa/crypto.h:3774:0,
                 from ../include/mbedtls/pk.h:49,
                 from pk.c:29:
pk.c: In function 'mbedtls_pk_wrap_as_opaque':
../include/psa/crypto_struct.h:460:33: error: 'bits' may be used uninitialized in this function [-Werror=maybe-uninitialized]
         attributes->core.bits = (psa_key_bits_t) bits;
                                 ^~~~~~~~~~~~~~~~~~~~~
pk.c:608:12: note: 'bits' was declared here
     size_t bits;
            ^~~~
```
I think setting `*bits = 0` in the `default:` leg of the `switch` statement of the function that was made inline would resolve this.

Otherwise this looks good to me, but as noted in the PR description this changes the ABI - not a problem, but something we need to remember when preparing the next release.

**Comment by Patater** [comment]
Rebased to set `*bits = 0` in mbedtls_ecc_group_to_psa(), by adding a commit sharing justification for the change.

**Comment by Patater** [comment]
The only CI failure is with the ABI test, which this PR changes intentionally. SO version will need bumping.

```
+ scripts/abi_check.py -o FETCH_HEAD -n HEAD -s identifiers --brief
Compatibility issues found for libmbedcrypto
<reports>
<report kind="binary" version="1.2">

<removed_symbols>
  <header name="psa_crypto.c">
    <library name="libmbedcrypto.so.4">
      <name>mbedtls_ecc_group_to_psa</name>
    </library>
  </header>
</removed_symbols>

</report>
<report kind="source" version="1.2">

<removed_symbols>
  <header name="psa_crypto.c">
    <library name="libmbedcrypto.so.4">
      <name>mbedtls_ecc_group_to_psa</name>
    </library>
  </header>
</removed_symbols>

</report>
</reports>
Checking evolution from FETCH_HEAD (d9f694960f10c23cc5d7cb6de440fe3584ff7979) to HEAD (91c06de53602a64063f1ae43d3a44a2562deaab9)
No compatibility issues for libmbedx509
No compatibility issues for libmbedtls
script returned exit code 1
```

**Comment by Patater** [comment]
Rebased to revise commit message. It said "the following" a bit too often.

File: include/psa/crypto_extra.h [hunk 1]
```diff
@@ -578,8 +578,55 @@ psa_status_t psa_get_key_domain_parameters(
  *                      (`PSA_ECC_CURVE_xxx`).
  * \return              \c 0 on failure (\p grpid is not recognized).
  */
-psa_ecc_curve_t mbedtls_ecc_group_to_psa( mbedtls_ecp_group_id grpid,
-                                          size_t *bits );
+static inline psa_ecc_curve_t mbedtls_ecc_group_to_psa( mbedtls_ecp_group_id grpid,
```
Comment (ronald-cron-arm) [inline include/psa/crypto_extra.h:581]: This is not a small function thus is it ok to inline it regarding code size? No other way to do that?

Comment (Patater) [reply include/psa/crypto_extra.h:581]: I'm not sure it is so large, but I'm not one to decide. The compiler may decide it is too large or complicated and avoid inlining, when code size optimizing flags are employed.

Alternatives would include making a new C file to implement this function, included in the S target build and again in the NS target build.

Comment (mpg) [reply include/psa/crypto_extra.h:581]: I decided to do some measurements:
```sh
scripts/config.pl baremetal
make CC=arm-none-eabi-gcc CFLAGS='-Os -mthumb -march=armv6-m' lib
arm-none-eabi-size -t library/libmbedcrypto.a | tail -n1
```
Before this PR: 236581
After this PR: 236601
Difference: +20 bytes

IMO this difference is perfectly acceptable, especially considering that this function is meant to disappear at some point, as it's an artifact of the transition from the legacy crypto APIs to PSA (should probably become useless at some point in the 3.x line and then be removed in 4.0).

**Comment by mpg** [comment]
Note: the only test failing in the CI is the API/ABI checker, which is fully expected.

File: include/psa/crypto_extra.h (hunk 1 above)
Comment (ronald-cron-arm) [reply include/psa/crypto_extra.h:581]: Given the small size impact and that it is a temporary situation, fine by me then.

**Pull request merged**